### PR TITLE
Replacing `hal.ex.submit_and_wait` with queued operations and fences.

### PR DESCRIPTION
**Pull request by benvanik**
Adding compiler/runtime support for lowering the asynchronous stream dialect ops into HAL ops, materializing a timeline (today just one but multiple in the future), and passing through to the runtime HAL module. This allowed for the removal of the existing placeholder submit_and_wait op and enables queue-ordered allocations to be implemented in the HAL.

This is likely not the final design but unblocks work on coroutines, queue-ordered allocations, webgpu, and plumbing fences through the user-facing API/native ABI. Future refinements may create overrides that use semaphores instead of fences to avoid fence heap allocations when not required, but for most single-function classic ML models once we plumb fences through the ABI no internal fences are required. The current timeline materialization also strictly orders all invocations where instead we should be able to elide those when there's no internal program state to protect.

Because the various HAL backends all need work (CUDA/ROCM in particular need massive work) nearly everything is synchronized exactly as it was before but now that synchronization happens in the IR we emit and we can selectively start supporting async per target.

Progress on #1285 (just need to put fences on the ABI!).
Progress on #8093 (added yieldable fence waits).
Progress on #9572 (added compiler/runtime glue for queue-ordered allocs).

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Stream/Transforms/test/elide_timepoints.mlir [hunk 1]
```diff
@@ -0,0 +1,84 @@
+// RUN: iree-opt --split-input-file --iree-stream-elide-timepoints %s | FileCheck %s
+
+// Tests that an immediate timepoint passed along a call edge is propagated.
+
+// DO NOT SUBMIT
+// function call/return
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/Stream/Transforms/test/elide_timepoints.mlir:6]: you have a few `DO NOT SUBMIT` tests in this file

**Review by ScottTodd — APPROVED** [review]
Awesome :)

File: compiler/src/iree/compiler/Dialect/Stream/Transforms/test/elide_timepoints.mlir (hunk 1 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/Stream/Transforms/test/elide_timepoints.mlir:6]: crap, forgot that I was procrastinating on these tests :P good catch!

**Comment by benvanik** [comment]
I reworked the elide timepoints pass in [Adding --iree-stream-elide-timepoints pass.](https://github.com/iree-org/iree/pull/9754/commits/b356e59706a3a8b2ed73d12b3f7102209ae673ee) to be simpler/support more cases (and have tests) - no major changes anywhere else and it's just simpler now (replaced nasty flaky fixed-point iteration algo with a single loop).

**Pull request merged**